core/dispatch_json.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+use crate::*;
+use futures::future::FutureExt;
+use serde_derive::Deserialize;
+use serde_json::json;
+use serde_json::Value;
+use std::future::Future;
+use std::pin::Pin;
+
+pub trait JsonError {
+  fn kind(&self) -> i32;
+  fn message(&self) -> String;
+}
+
+pub type JsonResult = Result<Value, Box<dyn JsonError>>;
+
+pub type AsyncJsonOp = Pin<Box<dyn Future<Output = JsonResult>>>;
+
+pub enum JsonOp {
+  Sync(Value),
+  Async(AsyncJsonOp),
+  /// AsyncUnref is the variation of Async, which doesn't block the program
+  /// exiting.
+  AsyncUnref(AsyncJsonOp),
+}
+
+fn serialize_result(promise_id: Option<u64>, result: JsonResult) -> Buf {
+  let value = match result {
+    Ok(v) => json!({ 
+      "ok": v, 
+      "promiseId": promise_id 
+    }),
+    Err(err) => json!({ 
+      "err": { 
+        "kind": err.kind(), 
+        "message": err.message() 
+      }, 
+      "promiseId": promise_id 
+    }),
+  };
+  serde_json::to_vec(&value).unwrap().into_boxed_slice()
+}
+
+impl JsonError for serde_json::Error {
+  fn kind(&self) -> i32 {
+    -1
+  }
+
+  fn message(&self) -> String {
+    self.to_string()
+  }
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct AsyncArgs {
+  promise_id: Option<u64>,
+}
+
+pub fn json_op<D>(d: D) -> impl Fn(&[u8], Option<ZeroCopyBuf>) -> CoreOp
+where
+  D: Fn(Value, Option<ZeroCopyBuf>) -> Result<JsonOp, Box<dyn JsonError>>,
+{
+  move |control: &[u8], zero_copy: Option<ZeroCopyBuf>| {
+    let async_args: AsyncArgs = match serde_json::from_slice(control) {
```

Comment:
I'm pretty sure failing to parse message should be "internal" error not using public error forwarding mechanism (so return `parseErr` instead of `err`?)